pycrumbs/layout.py

```diff
@@ -0,0 +1,165 @@
+
+        # Left/right anchors prevent nodes from floating too far beyond the boundaries of the tree
+        left_anchor = Dimensions(None, None, None, None, None, min_x, 0)
+        right_anchor = Dimensions(None, None, None, None, None, max_x, 0)
+
+        dims_by_uid = {dim.node.uid: dim for dim in dims}
+
+        dims = sorted(dims, key=lambda dim: dim.node_x)
+        for i in np.random.permutation(len(dims)):
+            left = dims[i - 1] if i > 0 else left_anchor
+            right = dims[i + 1] if i < len(dims) - 1 else right_anchor
+
+            # Force due to left neighbor
+            force_vec = spring_force((left.node_x, left.node_y) if left else None,
+                                     (dims[i].node_x, dims[i].node_y),
+                                     spring_constant=spring_constant,
+                                     neutral_length=neutral_length)
+
+            # Force due to right neighbor
+            force_vec += spring_force((right.node_x, right.node_y) if right else None,
+                                      (dims[i].node_x, dims[i].node_y),
+                                      spring_constant=spring_constant,
+                                      neutral_length=neutral_length)
+
+            # Force due to parent
+            parent = dims_by_uid.get(dims[i].node.parent.uid) if dims[i].node.parent else None
+            force_vec += parent_multiplier * spring_force((parent.node_x, parent.node_y) if parent else None,
+                                                          (dims[i].node_x, dims[i].node_y),
+                                                          spring_constant=spring_constant,
+                                                          neutral_length=neutral_length)
+
+            # We put the nodes on rails in the x dimension, i.e. ignore the y component of the force
+            dx = force_vec[0] * alpha
+            dims[i].x += dx
+            dims[i].node_x += dx
+
+    # Collect nodes by depth
+    dim_by_depth = defaultdict(list)
+    for dim in dimensions:
+        dim_by_depth[dim.node.depth].append(dim)
+
+    root_dim, = dim_by_depth[0]
+    min_x = root_dim.x
+    max_x = root_dim.x + root_dim.width
+
+    for _ in range(iterations):
+        for _, dims in dim_by_depth.items():
+            adjust_at_depth(dims, min_x, max_x)
+
+    return dimensions
+
+
+def tree_layout(root):
+    """\
+    Performs tree layout with non-overlapping bounding boxes.
+
+    :param root: root of the tree
+    :return: list of dimensions
+    """
+    def shift_all(dimensions, dx, dy):
+        """Shifts all dimnensions by the same amount"""
+        return [Dimensions(dim.node, dim.x + dx, dim.y + dy, dim.width, dim.height,
+                           node_x=dim.node_x + dx, node_y=dim.node_y + dy)
+                for dim in dimensions]
+
+    def bounding_box(dimensions):
+        """Computers the total dimensions (bounding box) of all dimensions in a list"""
```

Comment:
Super minor.  "Computers" is probable a typo.

Reply:
Grate catch, thanks